index/segment/mem/searcher.go

```diff
@@ -0,0 +1,176 @@
+func newSequentialSearcher(
+	q queryable,
+	negationMergeFn negationMergeFn,
+	pool segment.PostingsListPool,
+) searcher {
+	return &sequentialSearcher{
+		queryable:       q,
+		negationMergeFn: negationMergeFn,
+		pool:            pool,
+	}
+}
+
+func (s *sequentialSearcher) Query(query segment.Query) (
+	candidateDocIDs segment.PostingsList,
+	pendingFilterFn matchPredicate,
+	err error,
+) {
+	if err := validateQuery(query); err != nil {
+		return nil, nil, err
+	}
+
+	// order filters to ensure the first filter has no-negation
+	filters := orderFiltersByNonNegated(query.Filters)
+	sort.Sort(filters)
+
+	var (
+		candidateDocIds segment.PostingsList
+		predicates      = make([]matchPredicate, 0, len(query.Filters))
+	)
+	// TODO: support parallel fetching across segments/filters
+	for filterIdx, filter := range query.Filters {
+		if filterIdx == 0 && filter.Negate {
+			return nil, nil, errFirstFilterMustNotBeNegation
+		}
+
+		fetchedIds, pred, err := s.queryable.Filter(filter)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		if pred != nil {
+			predicates = append(predicates, pred)
+		}
+
+		// i.e. we don't have any documents for the given filter, can early terminate entire fn
+		if fetchedIds == nil {
+			return nil, nil, nil
+		}
+
+		if candidateDocIds == nil {
+			candidateDocIds = fetchedIds
+			continue
+		}
+
+		// TODO: evaluate perf impact of retrieving all candidate docIDs, waiting till end,
+		// sorting by size and then doing the intersection
+
+		// update candidate set
+		if filter.Negate {
+			s.negationMergeFn(candidateDocIds, fetchedIds)
+		} else {
+			candidateDocIds.Intersect(fetchedIds)
+		}
+
+		// early terminate if we don't have any docs in candidate set
+		if candidateDocIds.IsEmpty() {
+			return nil, nil, nil
+		}
+	}
+
+	// TODO: once we support multiple segments, we'll have to merge results
+	return candidateDocIds, matchPredicates(predicates).Fn(), nil
+}
+
+type matchPredicates []matchPredicate
+
+func (m matchPredicates) Fn() matchPredicate {
+	return func(d doc.Document) bool {
+		for _, fn := range m {
+			if !fn(d) {
+				return false
+			}
+		}
+		return true
+	}
+}
+
+// orderFiltersByNonNegated orders filters which are not negated first in the list.
+type orderFiltersByNonNegated []segment.Filter
+
+func (of orderFiltersByNonNegated) Len() int           { return len(of) }
+func (of orderFiltersByNonNegated) Swap(i, j int)      { of[i], of[j] = of[j], of[i] }
+func (of orderFiltersByNonNegated) Less(i, j int) bool { return !of[i].Negate && of[j].Negate }
+
+// validate any assumptions we have about queries.
+func validateQuery(q segment.Query) error {
+	// assuming we only support AndConjuctions for now.
+	if q.Conjunction != segment.AndConjunction {
+		return fmt.Errorf("query: %v has an invalid conjuction: %v", q, q.Conjunction)
+	}
+
+	// ensure query level have at-least one filter or sub-query
+	if len(q.Filters) == 0 && len(q.SubQueries) == 0 {
+		return fmt.Errorf("empty query specified")
```

Comment:
nit: Declare err with a `errNoQuery` instead of fmt.Errorf since its a constant string.

Reply:
sure thing